File: java/src/algorithms/roy_floyd/RoyFloyd.java
package algorithms.roy_floyd;

import java.io.BufferedReader;
import java.io.FileReader;
import java.util.StringTokenizer;

/**
 * @author Sebastian Claici
 */
public class RoyFloyd {

    /**
     * Representation of an undirected graph.
     */
    public static class Graph {
        private final boolean[][] adjacencyMatrix;
        private final int[][] costMatrix;

        private final int size;

        /**
         * Creates a new Graph object with sizes nodes. Each graph is
         * represented by the adjacency matrix (whether two nodes are
         * connected), and the cost matrix (the cost to travel between
         * two nodes).
         *
         * @param size Number of nodes in the graph.
         */
        public Graph(int size) {
            this.size = size;

            adjacencyMatrix = new boolean[size][size];
            costMatrix = new int[size][size];
        }

        /**
         * Add an edge between two nodes of certain cost. Because
         * the graph is undirected the effect is to add two edges:
         * one between src and dst, and one between dst and src.
         *
         * @param src Source node.
         * @param dst Destination node.
         * @param cost Cost of the edge.
         */
        public void addEdge(int src, int dst, int cost) {
            adjacencyMatrix[src][dst] = true;
            adjacencyMatrix[dst][src] = true;
            costMatrix[src][dst] = cost;
            costMatrix[dst][src] = cost;
        }

        /**
         * Computes the all-sources shortest paths of this graph.
         *
         * @return An array representing the cost matrix of the shortest paths.
         */
        public int[][] royFloyd() {
            int[][] result = new int[size][size];
            for (int i = 0; i < size; ++i)
                for (int j = 0; j < size; ++j)
                    result[i][j] = costMatrix[i][j];

            //Try to find a better path for each pair (i, j) through an auxiliary node k.
            for (int k = 0; k < size; ++k)
                for (int i = 0; i < size; ++i)
                    for (int j = 0; j < size; ++j) {
                        if (!adjacencyMatrix[i][k] || !adjacencyMatrix[k][j] || i == j)
                            continue;
                        result[i][j] = Math.min(result[i][j], result[i][k] + result[k][j]);
                    }

            return result;
        }
    }

    // Test code
    public static void main(String[] args) throws Exception {
        BufferedReader buf = new BufferedReader(new FileReader("royfloyd.in"));

        int N = Integer.parseInt(buf.readLine());
        Graph graph = new Graph(N);
        for (int i = 0; i < N; ++i) {
            StringTokenizer st = new StringTokenizer(buf.readLine());
            for (int j = 0; j < N; ++j) {
                int cost = Integer.parseInt(st.nextToken());
                graph.addEdge(i, j, cost);
            }
        }

        int[][] result = graph.royFloyd();
        for (int i = 0; i < N; ++i) {
            for (int j = 0; j < N; ++j) {
                System.out.printf("%d ", result[i][j]);
            }
            System.out.println();
        }
    }
}

File: cpp/algorithms/roy_floyd/roy_floyd.cpp
#include <fstream>
#include <iostream>
#include <limits>
#include <vector>

class graph {
 public:
    explicit graph(int _size)
        : size(_size), 
          cost_matrix(_size, std::vector<int>(_size)),
          adjacency_matrix(_size, std::vector<bool>(_size, false)) {}

    void add_edge(int src, int dst, int cost) {
        if (cost != 0)
            adjacency_matrix[src][dst] = true; 
        cost_matrix[src][dst] = cost;
    }

    std::vector<std::vector<int>> roy_floyd() {
        std::vector<std::vector<int>> result(size, std::vector<int>(size));

        for (auto i = 0; i < size; ++i)
            for (auto j = 0; j < size; ++j) {
                result[i][j] = cost_matrix[i][j];
                if (i != j && result[i][j] == 0)
                    result[i][j] = std::numeric_limits<int>::max() / 2 - 1;
            }

        for (auto k = 0; k < size; ++k) {
            for (auto i = 0; i < size; ++i) {
                for (auto j = 0; j < size; ++j) {
                    result[i][j] = std::min(result[i][j], 
                            result[i][k] + result[k][j]);
                }
            }
        }

        return result;
    }

 private:
    int size;
    std::vector<std::vector<int>> cost_matrix;
    std::vector<std::vector<bool>> adjacency_matrix;
};

int main()
{
    std::ifstream fin("royfloyd.in");
    std::ofstream fout("royfloyd.out");

    int N;
    fin >> N;

    graph g(N);
    for (auto i = 0; i < N; ++i)
        for (auto j = 0; j < N; ++j) {
            int cost;
            fin >> cost;
            g.add_edge(i, j, cost);
        }

    std::vector<std::vector<int>> result = g.roy_floyd();
    for (auto i = 0; i < result.size(); ++i) {
        for (auto j = 0; j < result[i].size(); ++j)
            fout << result[i][j] << " ";
        fout << "\n";
    }

    return 0;
}
